### PR TITLE
[FEATURE] Simple rendering of 3D linestrings

### PR DESCRIPTION
**Pull request by wonder-sk**
This mode of 3D line rendering will use OpenGL line rendering
instead of buffering lines into polygons and rendering them as meshes.

The advantage is that the 3D lines do not loose their Z coordinate
which is the case currently with "ordinary" 3D rendering after buffering.

The disadvantage is that the lines cannot be wide (supported in Qt3D only
since 5.10, but even then their rendering won't have nice joins/caps)
and only ambient color is used from the material.

image

(cc @nirvn aka joyful sunday continues!)

### CODE REVIEW

**Comment by nirvn** [comment]
Brilliant, may this Sunday be remember as a memorable one! :)

**Comment by nyalldawson** [comment]
@wonder-sk 

> The disadvantage is that the lines cannot be wide (supported in Qt3D only
since 5.10, but even then their rendering won't have nice joins/caps)

Is this worth exposing but hiding behind a version check? 

I'm also wondering if we should disable/hide the existing broken functionality for draping 3d lines?

File: src/3d/symbols/qgsline3dsymbol_p.cpp [hunk 1]
```diff
@@ -150,4 +154,105 @@ Qt3DRender::QGeometryRenderer *QgsLine3DSymbolEntityNode::renderer( const Qgs3DM
...
+  {
+    if ( f.geometry().isNull() )
+      continue;
+
+    if ( symbol.altitudeBinding() == AltBindCentroid )
+      centroid = QgsPoint( f.geometry().centroid().asPoint() );
+
+    QgsGeometry geom = f.geometry();
+    const QgsAbstractGeometry *g = geom.constGet();
+    if ( QgsLineString *ls = qgsgeometry_cast<QgsLineString *>( g ) )
```
Comment (nyalldawson) [inline src/3d/symbols/qgsline3dsymbol_p.cpp]: shouldn't this be 

    if ( const QgsLineString *ls = qgsgeometry_cast< const QgsLineString *>( g ) )

(Does qgsgeometry_cast allow casting away const? (it's a bug if it does). And if not, this cast would/should fail?)

File: src/3d/symbols/qgsline3dsymbol_p.cpp [hunk 2]
```diff
@@ -150,4 +154,105 @@ Qt3DRender::QGeometryRenderer *QgsLine3DSymbolEntityNode::renderer( const Qgs3DM
...
+    {
+      for ( int i = 0; i < ls->vertexCount(); ++i )
+      {
+        QgsPoint p = ls->vertexAt( QgsVertexId( 0, 0, i ) );
```
Comment (nyalldawson) [inline src/3d/symbols/qgsline3dsymbol_p.cpp]: QgsPoint p = ls->pointN( i );

is a bit more efficient

**Comment by nirvn** [comment]
I would also favor adding the size option for qt 5.10 and up.

**Comment by nirvn** [comment]
Also, as discussed over chat yesterday, I think we should make this rendering mode the default one. IMHO, mostly people would expect the Z value to be respected when activating the 3D rendering.

**Comment by NathanW2** [comment]
+1 on default

**Comment by nirvn** [comment]
@wonder-sk , there's a problem with the Y location of the simple 3D line:
fine

It looks like it's flipped (i.e. the distance between the northern edge and the beginning of the wrongly located 3D lines == the distance between the southern edge and the beginning of the correctly drawn 2D lines.

File: src/3d/symbols/qgsline3dsymbol_p.cpp [hunk 3]
```diff
@@ -150,4 +154,105 @@ Qt3DRender::QGeometryRenderer *QgsLine3DSymbolEntityNode::renderer( const Qgs3DM
...
+        float z = Qgs3DUtils::clampAltitude( p, symbol.altitudeClamping(), symbol.altitudeBinding(), symbol.height(), centroid, map );
+        vertices << QVector3D( p.x() - map.origin().x(), z, p.y() - map.origin().y() );
```
Comment (nirvn) [inline src/3d/symbols/qgsline3dsymbol_p.cpp]: On my machine, this leads to lines wrongly placed. The lines are properly placed when I changed y bit to `map.origin().y() - p.y()`.

fixed

File: src/3d/symbols/qgsline3dsymbol_p.cpp (hunk 1 above)
Comment (wonder-sk) [reply src/3d/symbols/qgsline3dsymbol_p.cpp]: It seems that's by design of qgsgeometry_cast - its definition includes const_cast - which probably should not be there. Changing this to const LineString...

**Comment by wonder-sk** [comment]
Thanks @nyalldawson and @nirvn for the reviews.

Regarding your points raised:

1. line width - let's address that in a future PR. I have two reasons: 1) I don't have currently QGIS running with Qt 5.10, 2) it is non-trivial amount of work: the line width is a part of the global OpenGL renderer state, which means we would need to have extra branches of the rendering frame graph for different line widths... it gets a bit complicated and the wide lines drawn by OpenGL looks quite bad (as mentioned - not line joins/caps, no anti-aliasing). I think both this approach and the older approach with lines buffered to polygons are just patches before we have a proper rendering code for lines. Something like this: https://www.codeproject.com/Articles/226569/Drawing-polylines-by-tessellation

2. make it default rendering style for linestrings - agreed it is probably better than the "buffered" approach. Let's make that true in a follow up PR...

**Comment by nirvn** [comment]
@wonder-sk , sound logic. +1 to merge on my side.

**Comment by nyalldawson** [comment]
> line width - let's address that in a future PR

Fair enough! I was under the mistaken assumption it was as simple as calling just another setter :stuck_out_tongue:

File: src/3d/symbols/qgsline3dsymbol_p.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/3d/symbols/qgsline3dsymbol_p.cpp]: @m-kuhn any feedback here? Can we remove the const_cast?

Comment (m-kuhn) [reply src/3d/symbols/qgsline3dsymbol_p.cpp]: Go ahead. I think the compiler will tell you quickly what the reason for introducing this was (if any) and will tell you at the same speed if alternative approaches do work or not.

**Pull request merged**